release-20.1: sql: wait for canceled mutations to be at the head of the queue for reverse

ajwerner opened the pull request:
Backport 1/1 commits from #54876.

/cc @cockroachdb/release

---

Before this commit, we'd totally skip the process of rolling back a mutation if
it were not the first mutation in line. This behavior was nonsense. Perhaps
there is a more principled change to be made whereby we just remove mutations
which are not at the head of the queue. I'm open to making such a change. This
seemed more pragmatic, though it may have problems I am not aware of.

Fixes #54861

Release note (bug fix): Fixed a bug which can occur when canceling schema
changes when there are multiple queued schema changes that could result in
future schema changes being stuck.

Commit:
sql: wait for canceled mutations to be at the head of the queue for rollback

Before this commit, we'd totally skip the process of rolling back a mutation if
it were not the first mutation in line. This behavior was nonsense. Perhaps
there is a more principled change to be made whereby we just remove mutations
which are not at the head of the queue. I'm open to making such a change. This
seemed more pragmatic, though it may have problems I am not aware of.

Fixes #54861

Release note (bug fix): Fixed a bug which can occur when canceling schema
changes when there are multiple queued schema changes that could result in
future schema changes being stuck.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55058)
<!-- Reviewable:end -->

thoszhang reviewed (approved):
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55058)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

Commit:
sql: properly handle cases where marking a job as canceled fails

In #55056 an assertion was added that was too strict. It is possible for a
revert to complete and then for the writing of the completed status to fail
and run again after the descriptor has had the mutation removed. This PR
allows that missing mutation to pass through just fine.

Release note (bug fix): Fix a bug which can lead to canceled schema change jobs
ending in the failed rather than canceled state.

Commit:
sql: clean up some cruft related to InvalidMutationID in maybeReverseMutations

Release note: None

ajwerner commented:
Needs #55513 too.

lucy-zhang commented:
Is this blocked on anything? It would be good to get it in for 20.1.10 next week.

ajwerner commented:
Just a nasty rebase. Do you know when the backport deadline for that release is?

lucy-zhang commented:
Monday night (the 14th), as usual.

ajwerner commented:
This is updated and RFAL

thoszhang reviewed (commented):
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55058)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schema_changer.go, line 1244 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55058#-MOXFS0S7d-e9wYVdjbq:-MOXFS0S7d-e9wYVdjbr:bc2prhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2728be096661703f12308955c9fa27611464e59/pkg/sql/schema_changer.go#L1244)):*
> ```Go
> 				// mutation ID we're looking for.
> 				if i == 0 {
> 					return nil
> ```

Does this change make a difference?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55058)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/schema_changer.go, line 1244 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55058#-MOXFS0S7d-e9wYVdjbq:-MOYQ84S-oZbmyagm2sv:b8ijee4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2728be096661703f12308955c9fa27611464e59/pkg/sql/schema_changer.go#L1244)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Does this change make a difference?
</blockquote></details>

This is wrong. This was just bad backporting on my part. updated. Thanks


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55058)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
